### PR TITLE
Document flexible controllers...

### LINKED ISSUE #168: Documentation for flexible controllers
From a discussion with @martin-drhu-da about flexible controllers. Here's what I think needs doing, docs-wise:

- [Reference: templates - choices](https://docs.daml.com/daml/reference/templates.html#choices) needs to show both the new and the old syntax in the code snippet, and have a bullet point explaining that there are two options for how you specify choices. This is just an overview page: the main information goes in...
- [Reference: choices](https://docs.daml.com/daml/reference/choices.html). This page will need a bit of restructuring, but I don't think too much. Essentially needs adjusting so it's clear that there are two different options of expressing this - the old way if you only need "normal" controllers, and the new way if you need flexible ones. Include the caveat about controllers not being added as observers automatically, and so you should sort this out to avoid messing yourself up at runtime. 

    The trick will be working out how to avoid duplicating information too much. Happy to talk to whoever does this work about how to solve this problem. The main thing I want to make sure is that the headings still provide a good overview of the page.
- [Multiple party agreement pattern](https://docs.daml.com/daml/patterns/multiparty-agreement.html) needs updating. I'm already on this. (@bethaitman-da)
- [Overview: template structure](https://docs.daml.com/daml/reference/structure.html) should show the new syntax as an option.

That's everything I can see that definitely needs changing. It might also be good to go through our public examples and see if any of them would benefit from the new feature.

### PR DESCRIPTION
**Pull request by bethaitman**
Fixes #168.

In this PR, I've also made some small improvements to the reference docs as I went, and moved all the snippets to using `start-after` and `end-before`, because otherwise updating them would have been a nightmare.

### CODE REVIEW

**Comment by bethaitman** [comment]
@martin-drhu-da FYI - hoping to get a decent draft of this tomorrow.

File: docs/source/daml/reference/structure.rst [hunk 1]
```diff
@@ -53,7 +51,11 @@ template body
         Only create the contract if the conditions after ``ensure`` evaluate to true.
 
     :ref:`choices <daml-ref-choice-structure>`
-    	``controller nameOfParty can nameOfChoice``
+    	``controller nameOfParty can nameOfChoice do``
+
+        or
+
+        ``choice : () controller do``
```
Comment (hurryabit) [inline docs/source/daml/reference/structure.rst]: To me this somehow indicates that the return type is always `()`, but it can be anything. How about
```
choice NameOfChoice : ReturnType controller nameOfParty do
```
If we do this, we should probably change the summary of the old syntax to
```
controller nameOfParty can NameOfChoice : ReturnType do
```
There's also the option to have choice arguments following the `with` keyword but I guess we don't want that here since we don't have it for the old syntax.

File: docs/source/daml/reference/structure.rst [hunk 2]
```diff
@@ -88,6 +94,8 @@ Here's the structure of a choice inside a template:
 :ref:`choice arguments <daml-ref-choice-arguments>`
     ``with`` keyword
 
+    If you start your choice with ``choice`` and include a ``Party`` as a parameter, you can make that ``Party`` the ``controller`` of the choice. This is a feature called "flexible controllers", and it means you don't have to specify the controller when you create the contract - you can specify it when you exercise the choice. The ``Party`` still needs to be a parameter to the contract as well, so that they can be added as an observer.
```
Comment (hurryabit) [inline docs/source/daml/reference/structure.rst]: I think the last sentence is slightly misleading. Maybe
> To exercise a choice, the party needs to be a signatory or an observer of the contract and must be explicitly declared as such.

File: docs/source/daml/reference/templates.rst [hunk 3]
```diff
@@ -69,15 +66,14 @@ Observers
...
+   :end-before: -- end template obs snippet
 
 - ``observer`` keyword. After ``where``. Followed by at least one ``Party``.
 - Observers are additional stakeholders, so the contract instance is visible to these parties (see the ``Party`` type).
-- Optional. You can have many, either as a comma-separated list or reusing the keyword.
+- Optional. You can have many, either as a comma-separated list or reusing the keyword. You could pass in a list (of type ``[Party]``).
 - Use when a party needs visibility on a contract, or be informed or contract events, but is not a :ref:`signatory <daml-ref-signatories>` or :ref:`controller <daml-ref-controllers>`.
-
-.. TODO: what about observer observers?
+- If you start your choice with ``choice`` rather than ``controller`` (see :ref:`daml-ref-choices` below), you must make sure to add the controller as an observer. Otherwise, they will not be able to exercise the choice, because they won't be able to see the contract.
```
Comment (hurryabit) [inline docs/source/daml/reference/templates.rst]: We should probably rephrase this as
> you must make sure to any potential controller as an observer
